solver/llbsolver/history.go

```diff
@@ -74,62 +322,188 @@ func (h *HistoryQueue) Update(ctx context.Context, e *controlapi.BuildHistoryEve
+	if err != nil {
+		return nil, err
+	}
+
+	return &Writer{
+		mt:    mt,
+		lm:    h.LeaseManager,
+		l:     l,
+		w:     w,
+		dgstr: digest.Canonical.Digester(),
+	}, nil
+}
+
+type Writer struct {
+	mt string
+	w  content.Writer
+	lm leases.Manager
+	l  leases.Lease
+
+	dgstr digest.Digester
+	sz    int
+}
+
+func (w *Writer) Write(p []byte) (int, error) {
+	if _, err := w.dgstr.Hash().Write(p); err != nil {
+		return 0, err
+	}
+	w.sz += len(p)
+	return w.w.Write(p)
+}
+
+func (w *Writer) Discard() {
+	w.w.Close()
+	w.lm.Delete(context.TODO(), w.l)
+}
+
+func (w *Writer) Commit(ctx context.Context) (*ocispecs.Descriptor, func(), error) {
+	dgst := w.dgstr.Digest()
+	sz := int64(w.sz)
+	if err := w.w.Commit(ctx, int64(w.sz), dgst); err != nil {
+		if !errdefs.IsAlreadyExists(err) {
+			w.Discard()
+			return nil, nil, err
+		}
+	}
+	return &ocispecs.Descriptor{
+			MediaType: w.mt,
+			Digest:    dgst,
+			Size:      sz,
+		},
+		func() {
+			w.lm.Delete(context.TODO(), w.l)
+		}, nil
+}
+
+func (h *HistoryQueue) ImportStatus(ctx context.Context, ch chan *client.SolveStatus) (_ *ocispecs.Descriptor, _ func(), err error) {
+	defer func() {
+		if ch == nil {
+			return
+		}
+		for range ch {
+		}
+	}()
+
+	w, err := h.OpenBlobWriter(ctx, "application/vnd.buildkit.status.v0")
+	if err != nil {
+		return nil, nil, err
+	}
+
+	bufW := bufio.NewWriter(w)
+
+	defer func() {
+		if err != nil {
+			w.Discard()
+		}
+	}()
+
+	buf := make([]byte, 32*1024)
+	for st := range ch {
+		hdr := make([]byte, 4)
+		for _, pst := range st.Marshal() {
+			sz := pst.Size()
+			if len(buf) < sz {
+				buf = make([]byte, sz)
 			}
-			dt, err := e.Record.Marshal()
+			n, err := pst.MarshalTo(buf)
 			if err != nil {
-				return err
+				return nil, nil, err
 			}
-
-			l, err := h.LeaseManager.Create(ctx, leases.WithID(h.leaseID(e.Record.Ref)))
-			if err != nil {
-				return err
+			binary.LittleEndian.PutUint32(hdr, uint32(n))
+			if _, err := bufW.Write(hdr); err != nil {
+				return nil, nil, err
 			}
-
-			defer func() {
-				if err != nil {
-					h.LeaseManager.Delete(ctx, l)
-				}
-			}()
-
-			if err := h.addResource(ctx, l, e.Record.Logs); err != nil {
-				return err
+			if _, err := bufW.Write(buf[:n]); err != nil {
+				return nil, nil, err
 			}
-
-			return b.Put([]byte(e.Record.Ref), dt)
-		}); err != nil {
-			return err
 		}
-		h.ps.Send(e)
 	}
-	return nil
+	if err := bufW.Flush(); err != nil {
+		return nil, nil, err
+	}
+	return w.Commit(ctx)
 }
 
-func (h *HistoryQueue) Listen(ctx context.Context, ref string, active bool, f func(*controlapi.BuildHistoryEvent) error) error {
+func (h *HistoryQueue) Listen(ctx context.Context, req *controlapi.BuildHistoryRequest, f func(*controlapi.BuildHistoryEvent) error) error {
 	h.init()
 
 	h.mu.Lock()
 	sub := h.ps.Subscribe()
 	defer sub.close()
 
+	if req.Ref != "" {
+		if _, ok := h.deleted[req.Ref]; ok {
+			h.mu.Unlock()
+			return errors.Wrapf(os.ErrNotExist, "ref %s is deleted", req.Ref)
+		}
```

Comment:
It looks like `h.deleted` grows forever, ideally we'd have some way of clearing it after some time?